packages/next/src/export/worker.ts

```diff
@@ -335,9 +316,166 @@ async function exportPageImpl(
   }
 }
 
-export default async function exportPage(
+export async function exportPages(
+  input: ExportPagesInput
+): Promise<ExportPagesResult> {
+  const {
+    exportPathMap,
+    paths,
+    dir,
+    distDir,
+    outDir,
+    cacheHandler,
+    cacheMaxMemorySize,
+    fetchCacheKeyPrefix,
+    enabledDirectories,
+    pagesDataDir,
+    renderOpts,
+    nextConfig,
+    options,
+  } = input
+
+  // If the fetch cache was enabled, we need to create an incremental
+  // cache instance for this page.
+  const incrementalCache = await createIncrementalCache({
+    cacheHandler,
+    cacheMaxMemorySize,
+    fetchCacheKeyPrefix,
+    distDir,
+    dir,
+    enabledDirectories: enabledDirectories,
+    // skip writing to disk in minimal mode for now, pending some
+    // changes to better support it
+    flushToDisk: !hasNextSupport,
+  })
+
+  renderOpts.incrementalCache = incrementalCache
+
+  const maxConcurrency =
+    nextConfig.experimental.staticGenerationMaxConcurrency ?? 2
+  const results: ExportPagesResult = []
+
+  const exportPageWithRetry = async (path: string, maxAttempts: number) => {
+    const pathMap = exportPathMap[path]
+    const { page } = exportPathMap[path]
+    const pageKey = page !== path ? `${page}: ${path}` : path
+    let attempt = 0
+    let result
+
+    while (attempt < maxAttempts) {
+      try {
+        result = await Promise.race<ExportPageResult | undefined>([
+          exportPage({
+            path,
+            pathMap,
+            distDir,
+            outDir,
+            pagesDataDir,
+            renderOpts,
+            ampValidatorPath:
+              nextConfig.experimental.amp?.validator || undefined,
+            trailingSlash: nextConfig.trailingSlash,
+            serverRuntimeConfig: nextConfig.serverRuntimeConfig,
+            subFolders: nextConfig.trailingSlash && !options.buildExport,
+            buildExport: options.buildExport,
+            optimizeFonts: nextConfig.optimizeFonts,
+            optimizeCss: nextConfig.experimental.optimizeCss,
+            disableOptimizedLoading:
+              nextConfig.experimental.disableOptimizedLoading,
+            parentSpanId: input.parentSpanId,
+            httpAgentOptions: nextConfig.httpAgentOptions,
+            debugOutput: options.debugOutput,
+            enableExperimentalReact: needsExperimentalReact(nextConfig),
+          }),
+          // If exporting the page takes longer than the timeout, reject the promise.
+          new Promise((_, reject) => {
```

Comment:
One concern with only handling this inside the worker itself, if the thread is blocked for whatever reason we could fail to retry properly which would have previously worked having the retry handling in the parent process.

Maybe we can still have this timeout tracking in the parent process 🤔

Reply:
As discussed not a blocker but we could track progress and only retry from the parent the failed paths sent to the worker